### PR TITLE
Feature/users endpoint

### PR DESCRIPTION
**Pull request by Jipperism**

### CODE REVIEW

File: schema.graphql [hunk 1]
```diff
@@ -850,4 +856,19 @@ input StringSearchOptions {
...
-scalar UUID
\ No newline at end of file
+scalar UUID
+
+type User {
+  """The address of the user"""
+  address: String!
+
+  """The avatar of the user"""
+  avatar: String!
```
Comment (bitbeckers) [inline schema.graphql:866]: Avatar and display name can be undefined I assume?

File: src/controllers/UserController.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      this.setStatus(400);
+      return {
+        success: false,
+        message: "Invalid input",
+        data: null,
+        errors: JSON.parse(parsedBody.error.toString()),
+      };
+    }
+
+    const { signature, message } = parsedBody.data;
```
Comment (bitbeckers) [inline src/controllers/UserController.ts:53]: What is in the signed message? Does it contain a chainId? Based on line 54 it looks like this is fixed to Optimism, but we should be able to handle different chains without introducing breaking changes. This entail that the message should contain a chainId to a) prevent replays and b) support account abstraction methods like Safe

File: src/services/SupabaseDataService.ts [hunk 3]
```diff
@@ -197,4 +198,26 @@ export class SupabaseDataService {
       .eq("id", orderId)
       .single();
   }
+
+  async upsertUsers(
+    users: DataDatabase["public"]["Tables"]["users"]["Insert"][],
+  ) {
+    return this.supabaseData
```
Comment (bitbeckers) [inline src/services/SupabaseDataService.ts:205]: We discussed to start using the kysely client as a replacement for supabase, is there a specific reason why this is implemented using supabase?

Filtering, sorting and pagination are handled in the baseResolver defined in `src/graphql/schemas/resolvers/baseTypes.ts`

File: schema.graphql (hunk 1 above)
Comment (Jipperism) [reply schema.graphql]: Resolved in https://github.com/hypercerts-org/hypercerts-api/pull/168/commits/bd5b11c3a23a262de2d68bd710baaeaff3f98a5f

File: src/services/SupabaseDataService.ts (hunk 3 above)
Comment (Jipperism) [reply src/services/SupabaseDataService.ts]: Resolved in https://github.com/hypercerts-org/hypercerts-api/pull/168/commits/bd5b11c3a23a262de2d68bd710baaeaff3f98a5f

File: src/controllers/UserController.ts (hunk 2 above)
Comment (Jipperism) [reply src/controllers/UserController.ts:53]: To be clear, this never hits the chain so I'm not sure if replays are a thing (if I understand correctly)? Only if the request would be intercepted, you could change somebody profile info again but that would still be possible when adding a chain id to the message body 🤔 

How do you envision this. AFAICT there's 3 options:

- Sign a message like `${chainId} - ${address} - Update user settings`.
- Sign typed data such as `{ chainId: number; address: string; message: "Update user settings"
- Simply send chainId in the request as well and initialize the client with that.

Comment (Jipperism) [reply src/controllers/UserController.ts:53]: [Went through this comment and the links posted](https://discord.com/channels/1075404472152494100/1280134960212217898/1288039895750676553) and seems like typed data would be the option.

Comment (Jipperism) [reply src/controllers/UserController.ts]: Implemented this in https://github.com/hypercerts-org/hypercerts-api/pull/168/commits/29b5e9c348349660a3d2bf0dcb395672df1b64fa, @bitbeckers please let me know if you agree with the implementation.

**Comment by Jipperism** [comment]
This PR now also requires setting the `DATA_DATABASE_URL` env var in railway, for hypercerts-data-staging

File: src/controllers/UserController.ts (hunk 2 above)
Comment (bitbeckers) [reply src/controllers/UserController.ts]: Thanks. As it follows EIP712, it need the DomainSeparator as defined here: https://eips.ethereum.org/EIPS/eip-712#definition-of-domainseparator

In our case, only field `name, version, chainId` are applicable. This lead to having two fields names `chainId` and `address`, as the `signer` is the `address` and `chainId` is in the DomainSeparator.

Is there a case where we want address and chainId explicitly declared outside of the DomainSeparator?

File: src/controllers/UserController.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        name: "Hypercerts",
+        version: "1",
+        chainId: chain_id,
+      },
+      types: {
+        User: [
+          { name: "displayName", type: "string" },
+          { name: "avatar", type: "string" },
+        ],
+        UpdateRequest: [{ name: "user", type: "User" }],
```
[the remark this answers is not in the document]
Comment (Jipperism) [reply src/controllers/UserController.ts]: resolved

**Pull request merged**